core-bundle/src/EventListener/PreviewToolbarListener.php

```diff
@@ -0,0 +1,109 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\EventListener;
+
+use Contao\CoreBundle\Routing\ScopeMatcher;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Event\ResponseEvent;
+use Symfony\Component\Routing\RouterInterface;
+use Twig\Environment as TwigEnvironment;
+use Twig\Error\Error as TwigError;
+
+/**
+ * PreviewToolbarListener injects the back end preview toolbar on any response within the /preview.php entry point.
+ *
+ * The onKernelResponse method must be connected to the kernel.response event.
+ *
+ * The toolbar is only injected on well-formed HTML (with a proper <body> tag).
+ * This means that the toolbar is never included in sub-requests or ESI requests.
+ */
+class PreviewToolbarListener
+{
+    private $scopeMatcher;
+
+    private $previewScript;
+
+    private $twig;
+
+    private $router;
+
+    public function __construct(
+        string $previewScript,
+        ScopeMatcher $scopeMatcher,
+        TwigEnvironment $twig,
+        RouterInterface $router
+    ) {
+        $this->previewScript = $previewScript;
+        $this->scopeMatcher = $scopeMatcher;
+        $this->twig = $twig;
+        $this->router = $router;
+    }
+
+    public function onKernelResponse(ResponseEvent $event): void
+    {
+        if (!$this->scopeMatcher->isFrontendMasterRequest($event)) {
+            return;
+        }
+
+        $request = $event->getRequest();
+        $response = $event->getResponse();
+
+        if ($request->getScriptName() !== $this->previewScript) {
```

Comment:
Is this the only way to detect the preview script? Be aware that the script can be empty, and I would assume then we would not inject anything?

Reply:
What do you mean by "the script can be empty" and what do you propose?

Reply:
The preview script is configurable so in theory it would be possible to enter `''` there. However, this would require quite a bit of reverse proxy configuration (if it's not impossible at all) and I don't think it's a use case that really should be supported. Configuring it to `p.php` or whatever, yes. But leaving it empty...I don't know. I'm fairly confident this will never happen.

Reply:
it will happen, if you do not have a reverse proxy. Then you also do not need an entry point script…